OpenTrons/opentrons-api
# feat(api): turn off modules when estop is pressed and forward estop messages to api listeners

# Overview

When the estop is pressed, we need to deactivate the following module behaviors:
* Any thermal control
* Any shaking

This PR adds that automatic deactivation to the OT3Api class. In addition, this PR also hooks up the OT3Api to forward any EstopStateNotification events that occur as a result of the Estop being pressed, connected/disconnected, or acknowledged.

Note that, per the PRD, modules are not prevented from performing any thermal control once the Estop is engaged; they just need to be turned off when the estop is first pressed.

# Test Plan

On a Flex with the Thermocycler, Heater-Shaker, and Tempdeck I did the following:

* Set all of the modules to control temperature, and set the heater shaker to shake
* Hit the estop and observe all of the modules deactivating
* Repeat the first step and then start a protocol
* **Pause** the protocol and then hit the estop; observe all of the modules deactivating 

# Changelog
* Add ability to force module deactivation even if the execution manager is not running (protocol paused, etc)
* Add listener in OT3Api that is hooked up to the estop state machine to listen for (and forward) any estop events
* When an estop event signals that the estop was just pressed, deactivate all modules

# Review requests

I don't like the `asyncio.sleep` in the new test at all. The estop listener _needs_ to start the deactivation all at once and it _shouldn't_ block and wait for all of the deactivation tasks to complete though. 

# Risk assessment

Pretty low
